gobblin-api/src/main/java/gobblin/publisher/DataPublisher.java

```diff
@@ -18,19 +18,38 @@
 import java.lang.reflect.InvocationTargetException;
 import java.util.Collection;
 
+import com.google.common.base.Optional;
+import com.google.common.collect.Iterables;
+
+import gobblin.commit.CommitSequence;
+import gobblin.commit.DeliverySemantics;
+import gobblin.configuration.ConfigurationKeys;
 import gobblin.configuration.State;
 import gobblin.configuration.WorkUnitState;
+import lombok.Getter;
 
 
 /**
  * Defines how to publish data and its corresponding metadata. Can be used for either task level or job level publishing.
+ *
+ * If exactly-once semantics is used, it will construct a {@link CommitSequence} instead of actually publishing
+ * the data. The data will be published when the {@link CommitSequence} is executed, which will be after the
+ * {@link CommitSequence} has been fully constructed and persisted into the {@link gobblin.commit.CommitSequenceStore}.
  */
 public abstract class DataPublisher implements Closeable {
```

Comment:
I think for now we should use a different publisher class. I don't think the current API for DataPublisher fits into the new model. For example, there shouldn't be publishData() and publishMetadata() methods. All of these should be commit steps.

We'll have to figure out later how to do the migration. We may have to introduce some common interface.